### ISSUE TITLE
Installation error: `💥 maturin failed`

### ISSUE THREAD

**Issue by olgabot** [sourmash-bio/sourmash_plugin_branchwater#148]
Hello,
Hope you are doing well! I am excited to try this plugin :)

I am having installation issues. First, I cloned the repo and created a new environment using the provided `environment.yml`, which went fine.

<details><summary>Conda env creation</summary>

```
(sourmash-v4.8.4)
 ✘  Sat 11 Nov - 15:50  ~/code/pyo3_branchwater   origin ☊ main ✔ 
 @olgabot  conda env create --file environment.yml
Collecting package metadata (repodata.json): done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.14.0
  latest version: 23.10.0

Please update conda by running

    $ conda update -n base -c conda-forge conda



Downloading and Extracting Packages
maturin-1.3.1        | 4.3 MB    | ################################################################################################################### | 100%
rust-std-x86_64-appl | 29.3 MB   | ################################################################################################################### | 100%
pandas-2.1.3         | 13.4 MB   | ################################################################################################################### | 100%
rust-1.73.0          | 183.0 MB  | ################################################################################################################### | 100%
Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate branchwater-dev
#
# To deactivate an active environment, use
#
#     $ conda deactivate

Retrieving notices: ...working... done
(sourmash-v4.8.4)
 Sat 11 Nov - 15:58  ~/code/pyo3_branchwater   origin ☊ main ✔ 
 @olgabot  conda activate branchwater-dev
```

</details>

The `pip` log alone wasn't so informative:

<details><summary>`pip install -e .` log</summary>

```
Obtaining file:///Users/olgabot/code/pyo3_branchwater
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Checking if build backend supports build_editable: started
  Checking if build backend supports build_editable: finished with status 'done'
  Getting requirements to build editable: started
  Getting requirements to build editable: finished with status 'done'
  Preparing editable metadata (pyproject.toml): started
  Preparing editable metadata (pyproject.toml): finished with status 'done'
Requirement already satisfied: sourmash<5,>=4.8.3 in /Users/olgabot/opt/miniconda3/envs/branchwater-dev/lib/python3.12/site-packages (from pyo3-branchwater==0.8.2) (4.8.4)
Requirement already satisfied: screed<2,>=1.1.2 in /Users/olgabot/opt/miniconda3/envs/branchwater-dev/lib/python3.12/site-packages (from sourmash<5,>=4.8.3->pyo3-branchwater==0.8.2) (1.1.2)
Requirement already satisfied: cffi>=1.14.0 in /Users/olgabot/opt/miniconda3/envs/branchwater-dev/lib/python3.12/site-packages (from sourmash<5,>=4.8.3->pyo3-branchwater==0.8.2) (1.16.0)
Requirement already satisfied: numpy in /Users/olgabot/opt/miniconda3/envs/branchwater-dev/lib/python3.12/site-packages (from sourmash<5,>=4.8.3->pyo3-branchwater==0.8.2) (1.26.0)
Requirement already satisfied: matplotlib in /Users/olgabot/opt/miniconda3/envs/branchwater-dev/lib/python3.12/site-packages (from sourmash<5,>=4.8.3->pyo3-branchwater==0.8.2) (3.8.1)
Requirement already satisfied: scipy in /Users/olgabot/opt/miniconda3/envs/branchwater-dev/lib/python3.12/site-packages (from sourmash<5,>=4.8.3->pyo3-branchwater==0.8.2) (1.11.3)
Requirement already satisfied: deprecation>=2.0.6 in /Users/olgabot/opt/miniconda3/envs/branchwater-dev/lib/python3.12/site-packages (from sourmash<5,>=4.8.3->pyo3-branchwater==0.8.2) (2.1.0)
Requirement already satisfied: cachetools<6,>=4 in /Users/olgabot/opt/miniconda3/envs/branchwater-dev/lib/python3.12/site-packages (from sourmash<5,>=4.8.3->pyo3-branchwater==0.8.2) (4.2.4)
Requirement already satisfied: bitstring<5,>=3.1.9 in /Users/olgabot/opt/miniconda3/envs/branchwater-dev/lib/python3.12/site-packages (from sourmash<5,>=4.8.3->pyo3-branchwater==0.8.2) (3.1.9)
Requirement already satisfied: pycparser in /Users/olgabot/opt/miniconda3/envs/branchwater-dev/lib/python3.12/site-packages (from cffi>=1.14.0->sourmash<5,>=4.8.3->pyo3-branchwater==0.8.2) (2.21)
Requirement already satisfied: packaging in /Users/olgabot/opt/miniconda3/envs/branchwater-dev/lib/python3.12/site-packages (from deprecation>=2.0.6->sourmash<5,>=4.8.3->pyo3-branchwater==0.8.2) (23.2)
Requirement already satisfied: contourpy>=1.0.1 in /Users/olgabot/opt/miniconda3/envs/branchwater-dev/lib/python3.12/site-packages (from matplotlib->sourmash<5,>=4.8.3->pyo3-branchwater==0.8.2) (1.2.0)
Requirement already satisfied: cycler>=0.10 in /Users/olgabot/opt/miniconda3/envs/branchwater-dev/lib/python3.12/site-packages (from matplotlib->sourmash<5,>=4.8.3->pyo3-branchwater==0.8.2) (0.12.1)
Requirement already satisfied: fonttools>=4.22.0 in /Users/olgabot/opt/miniconda3/envs/branchwater-dev/lib/python3.12/site-packages (from matplotlib->sourmash<5,>=4.8.3->pyo3-branchwater==0.8.2) (4.44.0)
Requirement already satisfied: kiwisolver>=1.3.1 in /Users/olgabot/opt/miniconda3/envs/branchwater-dev/lib/python3.12/site-packages (from matplotlib->sourmash<5,>=4.8.3->pyo3-branchwater==0.8.2) (1.4.5)
Requirement already satisfied: pillow>=8 in /Users/olgabot/opt/miniconda3/envs/branchwater-dev/lib/python3.12/site-packages (from matplotlib->sourmash<5,>=4.8.3->pyo3-branchwater==0.8.2) (10.1.0)
Requirement already satisfied: pyparsing>=2.3.1 in /Users/olgabot/opt/miniconda3/envs/branchwater-dev/lib/python3.12/site-packages (from matplotlib->sourmash<5,>=4.8.3->pyo3-branchwater==0.8.2) (3.1.1)
Requirement already satisfied: python-dateutil>=2.7 in /Users/olgabot/opt/miniconda3/envs/branchwater-dev/lib/python3.12/site-packages (from matplotlib->sourmash<5,>=4.8.3->pyo3-branchwater==0.8.2) (2.8.2)
Requirement already satisfied: six>=1.5 in /Users/olgabot/opt/miniconda3/envs/branchwater-dev/lib/python3.12/site-packages (from python-dateutil>=2.7->matplotlib->sourmash<5,>=4.8.3->pyo3-branchwater==0.8.2) (1.16.0)
Building wheels for collected packages: pyo3-branchwater
  Building editable for pyo3-branchwater (pyproject.toml): started
  Building editable for pyo3-branchwater (pyproject.toml): still running...
  Building editable for pyo3-branchwater (pyproject.toml): finished with status 'error'
Failed to build pyo3-branchwater
```

</details>

When building on the command line, here is what I saw last:

<details><summary>The final part of the error message is:</summary>

```
        cargo:warning=1 error generated.

        exit status: 0
        exit status: 1
        cargo:warning=ToolExecError: Command "c++" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-m64" "-arch" "x86_64" "-I" "snappy/" "-I" "." "-Wall" "-std=c++11" "-DNDEBUG=1" "-o" "/Users/olgabot/code/pyo3_branchwater/target/release/build/librocksdb-sys-cd3040b71ec0d4d1/out/snappy/snappy.o" "-c" "snappy/snappy.cc" with args "c++" did not execute successfully (status code exit status: 1).
        exit status: 0

        --- stderr


        error occurred: Command "c++" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-m64" "-arch" "x86_64" "-I" "snappy/" "-I" "." "-Wall" "-std=c++11" "-DNDEBUG=1" "-o" "/Users/olgabot/code/pyo3_branchwater/target/release/build/librocksdb-sys-cd3040b71ec0d4d1/out/snappy/snappy.o" "-c" "snappy/snappy.cc" with args "c++" did not execute successfully (status code exit status: 1).


      💥 maturin failed
        Caused by: Failed to build a native library through cargo
        Caused by: Cargo build finished with "exit status: 101": `PYO3_ENVIRONMENT_SIGNATURE="cpython-3.12-64bit" PYO3_PYTHON="/Users/olgabot/opt/miniconda3/envs/branchwater-dev/bin/python3.12" PYTHON_SYS_EXECUTABLE="/Users/olgabot/opt/miniconda3/envs/branchwater-dev/bin/python3.12" "cargo" "rustc" "--message-format" "json-render-diagnostics" "--manifest-path" "/Users/olgabot/code/pyo3_branchwater/Cargo.toml" "--release" "--lib" "--" "-C" "link-arg=-undefined" "-C" "link-arg=dynamic_lookup" "-C" "link-args=-Wl,-install_name,@rpath/pyo3_branchwater.cpython-312-darwin.so"`
      Error: command ['maturin', 'pep517', 'build-wheel', '-i', '/Users/olgabot/opt/miniconda3/envs/branchwater-dev/bin/python3.12', '--compatibility', 'off', '--editable'] returned non-zero exit status 1
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building editable for pyo3-branchwater
ERROR: Could not build wheels for pyo3-branchwater, which is required to install pyproject.toml-based projects
```

</details>

<details><summary>Looking through the logs, this error message seems a bit more informative:</summary>

```
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building editable for pyo3-branchwater
Failed to build pyo3-branchwater
ERROR: Could not build wheels for pyo3-branchwater, which is required to install pyproject.toml-based projects
[1]  + 47344 exit 1     pip install -e .
  error: subprocess-exited-with-error

  × Building editable for pyo3-branchwater (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [830 lines of output]
      Running `maturin pep517 build-wheel -i /Users/olgabot/opt/miniconda3/envs/branchwater-dev/bin/python3.12 --compatibility off --editable`
      📦 Including license file "/Users/olgabot/code/pyo3_branchwater/LICENSE.txt"
      🍹 Building a mixed python/rust project
      🔗 Found pyo3 bindings
      🐍 Found CPython 3.12 at /Users/olgabot/opt/miniconda3/envs/branchwater-dev/bin/python3.12
          Blocking waiting for file lock on build directory
         Compiling librocksdb-sys v0.11.0+8.1.1
      The following warnings were emitted during compilation:

      warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/olgabot/code/pyo3_branchwater/target/release/build/librocksdb-sys-cd3040b71ec0d4d1/out/librocksdb.a(db_impl_debug.o) has no symbols
      warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/olgabot/code/pyo3_branchwater/target/release/build/librocksdb-sys-cd3040b71ec0d4d1/out/librocksdb.a(db_impl_debug.o) has no symbols
      warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/olgabot/code/pyo3_branchwater/target/release/build/librocksdb-sys-cd3040b71ec0d4d1/out/librocksdb.a(db_impl_debug.o) has no symbols
      warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/olgabot/code/pyo3_branchwater/target/release/build/librocksdb-sys-cd3040b71ec0d4d1/out/librocksdb.a(async_file_reader.o) has no symbols
      warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/olgabot/code/pyo3_branchwater/target/release/build/librocksdb-sys-cd3040b71ec0d4d1/out/librocksdb.a(crc32c_arm64.o) has no symbols
      warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/olgabot/code/pyo3_branchwater/target/release/build/librocksdb-sys-cd3040b71ec0d4d1/out/librocksdb.a(db_impl_debug.o) has no symbols
      warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/olgabot/code/pyo3_branchwater/target/release/build/librocksdb-sys-cd3040b71ec0d4d1/out/librocksdb.a(async_file_reader.o) has no symbols
      warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/olgabot/code/pyo3_branchwater/target/release/build/librocksdb-sys-cd3040b71ec0d4d1/out/librocksdb.a(crc32c_arm64.o) has no symbols
      warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/olgabot/code/pyo3_branchwater/target/release/build/librocksdb-sys-cd3040b71ec0d4d1/out/librocksdb.a(db_impl_debug.o) has no symbols
      warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/olgabot/code/pyo3_branchwater/target/release/build/librocksdb-sys-cd3040b71ec0d4d1/out/librocksdb.a(async_file_reader.o) has no symbols
      warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/olgabot/code/pyo3_branchwater/target/release/build/librocksdb-sys-cd3040b71ec0d4d1/out/librocksdb.a(crc32c_arm64.o) has no symbols
      warning: snappy/snappy.cc:1033:36: error: invalid output constraint '=@ccz' in asm
      warning:       : [tag_type] "+r"(tag_type), "=@ccz"(is_literal));
      warning:                                    ^
      warning: 1 error generated.
      warning: ToolExecError: Command "c++" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-m64" "-arch" "x86_64" "-I" "snappy/" "-I" "." "-Wall" "-std=c++11" "-DNDEBUG=1" "-o" "/Users/olgabot/code/pyo3_branchwater/target/release/build/librocksdb-sys-cd3040b71ec0d4d1/out/snappy/snappy.o" "-c" "snappy/snappy.cc" with args "c++" did not execute successfully (status code exit status: 1).
```

</details>

Do you know what may be happening?
Thank you so much!
Warmest,
Olga

**Comment by ctb** [comment]
hi @olgabot !!

I think @ccbaumler ran into something similar but I haven't tracked it down yet - try adding `compilers` to the environment and see if that helps?

**Comment by yuzie0314** [comment]
Hi @ctb, 
I just tried adding `compilers` to the environment but still got the same error posted by @olgabot.

**Comment by ccbaumler** [comment]
The issue referenced by @ctb 

https://github.com/sourmash-bio/pyo3_branchwater/issues/127

**Comment by ctb** [comment]
This is fixed by #151, I think.

> Hi @ctb, I just tried adding `compilers` to the environment but still got the same error posted by @olgabot.

@yuzie0314 thanks for checking it out - did you recreate the conda environment from the new environment.yml, or did you update the conda environment with `mamba install compilers`?

**Comment by ctb** [comment]
thanks @yuzie0314 - what platform/architecture are you on? What's the output of `uname -a`?
